add http2 ping health checks

tarat44 opened the pull request:
This pull request adds http2 ping health checks and would close #6260. Some tests are failing locally, but the errors seem unrelated.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/consul?pullRequest=8431)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/consul?pullRequest=8431)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/consul?pullRequest=8431).</sub>

tarat44 commented:
It looks like there is a bug with how the proxy is set (I have it using the same inject address function as http, but it shouldn't have https/http or slashes in the schema). I will fix this issue tonight.

tarat44 commented:
> It looks like there is a bug with how the proxy is set (I have it using the same inject address function as http, but it shouldn't have https/http or slashes in the schema). I will fix this issue tonight.

It actually appears that proxies do not support ping https://tools.ietf.org/html/rfc7540#section-8.3 so I will just remove all proxy references completely

tarat44 commented:
I just realized that the contributing check lists cover this change so I have made many of the missing changes and am pasting it below. I plan to update the documentation and give this list another look this weekend. I apologize for the unintended incompleteness of this pull request.

 - [x] Add the field to the Config struct (or an appropriate sub-struct) in
   `agent/config/config.go`.
 - [ ] Add the field to the actual RuntimeConfig struct in
   `agent/config/runtime.go`.
 - [x] Add an appropriate parser/setter in `agent/config/builder.go` to
   translate.
 - [x] Add the new field with a random value to both the JSON and HCL blobs in
   `TestFullConfig` in `agent/config/runtime_test.go`, it should fail now, then
   add the same random value to the expected struct in that test so it passes
   again.
 - [x] Add the new field and it's default value to `TestSanitize` in the same
   file. (Running the test first gives you a nice diff which can save working
   out where etc.)
 - [ ] **If** your new config field needed some validation as it's only valid in
   some cases or with some values (often true).
      - [ ] Add validation to Validate in `agent/config/builder.go`.
      - [ ] Add a test case to the table test `TestConfigFlagsAndEdgeCases` in
        `agent/config/runtime_test.go`.
 - [ ] **If** your new config field needs a non-zero-value default.
      - [ ] Add that to `DefaultSource` in `agent/config/defaults.go`.
      - [ ] Add a test case to the table test `TestConfigFlagsAndEdgeCases` in
        `agent/config/runtime_test.go`.
 - [ ] **If** your config should take effect on a reload/HUP.
      - [ ] Add necessary code to to trigger a safe (locked or atomic) update to
        any state the feature needs changing. This needs to be added to one or
        more of the following places:
         - `ReloadConfig` in `agent/agent.go` if it needs to affect the local
           client state or another client agent component.
         - `ReloadConfig` in `agent/consul/client.go` if it needs to affect
           state for client agent's RPC client.
      - [ ] Add a test to `agent/agent_test.go` similar to others with prefix
        `TestAgent_reloadConfig*`.
 - [ ] Add documentation to `website/source/docs/agent/options.html.md`.

tarat44 commented:
I have added documentation and another test. I think this pull request should be ready for review now

dnephin commented on agent/checks/check_test.go:
```diff
@@ -936,6 +937,115 @@ func TestCheckTCPPassing(t *testing.T) {
+	}
+}
+
+func TestCheckH2PINGInvalidListener(t *testing.T) {
```
An additional test case, similar to [https://github.com/hashicorp/consul/blob/bbc69636ec0553cc273692fd9cda07d1e3308987/agent/checks/check_test.go#L730-L770](TestCheckHTTP_TLS_BadVerify) would be good to verify that the correct TLS certs are being used.

dnephin reviewed (commented):
Thank you for the PR, this looks really good!

Sorry it took us so long to respond. Unfortunately there are a number of conflicts now that need to be resolved. 

I think once this is rebased to resolve the conflicts, and the additional test is added, this should be ready to merge.

tarat44 commented:
> Thank you for the PR, this looks really good!
> 
> Sorry it took us so long to respond. Unfortunately there are a number of conflicts now that need to be resolved.
> 
> I think once this is rebased to resolve the conflicts, and the additional test is added, this should be ready to merge.

Thank you for the review, I added the test for cert failure and rebased but still need to work on getting all of the tests to pass and catching up on syncing this new feature with more recent changes to the codebase. I hope to have time to complete the rest of this work sometime this week

tarat44 commented:
@dnephin  this pull request should be good to be looked over again. I hope it is okay that I added `H2PING` to `proto/pbservice/healthcheck.gen.go` manually as I was unsure how to generate it with`mod` as referenced in the comments. If you let me know how I can find and use this package, I would be happy to go back and do it correctly.

dnephin commented on agent/config/config.go:
```diff
@@ -411,6 +411,7 @@ type CheckDefinition struct {
 	AliasService                   *string             `mapstructure:"alias_service"`
 	Timeout                        *string             `mapstructure:"timeout"`
 	TTL                            *string             `mapstructure:"ttl"`
+	H2PING                         *string             `json:"h2ping,omitempty" hcl:"h2ping" mapstructure:"h2ping"`
```
```suggestion
	H2PING                         *string             `mapstructure:"h2ping"`
```

Minor: we removed the json and hcl struct tags here, since they haven't been used in some time.

dnephin commented on agent/checks/check_test.go:
```diff
@@ -964,6 +965,162 @@ func TestCheckTCPPassing(t *testing.T) {
+			check.Start()
+			defer check.Stop()
+
+			if tt.passing == true {
```
minor style note:

```suggestion
			if tt.passing {
```

dnephin reviewed (approved):
Thank you for implementing this change! Looks great!

I made a couple very minor suggestions. I'm going to apply those changes, and if the tests are still happy I think this is ready to merge.

I noticed that some of the checks use a WaitGroup to wait for the goroutine to stop. An example of adding that is 1910e2a246c0d8758c3c8af5891db4c342089abd. It's not strictly necessary, but can help prevent flaky tests and makes it easier to detect goroutine leaks in tests if we wait on ones we expect to shutdown. That might be a good addition to this check, as a follow up, if you are interested.

I also noticed the HTTP check supports a proxy. Would a proxy be appropriate for H2 checks as well? Not something we need to add right away, but maybe we could track adding that in a github issue in case it is useful for someone.

dnephin commented:
> I hope it is okay that I added H2PING to proto/pbservice/healthcheck.gen.go manually as I was unsure how to generate it with mog as referenced in the comments.

Yes, that perfect, thank you for making the change!

tarat44 commented:
> Thank you for implementing this change! Looks great!
> 
> I made a couple very minor suggestions. I'm going to apply those changes, and if the tests are still happy I think this is ready to merge.
> 
> I noticed that some of the checks use a WaitGroup to wait for the goroutine to stop. An example of adding that is [1910e2a](https://github.com/hashicorp/consul/commit/1910e2a246c0d8758c3c8af5891db4c342089abd). It's not strictly necessary, but can help prevent flaky tests and makes it easier to detect goroutine leaks in tests if we wait on ones we expect to shutdown. That might be a good addition to this check, as a follow up, if you are interested.
> 
> I also noticed the HTTP check supports a proxy. Would a proxy be appropriate for H2 checks as well? Not something we need to add right away, but maybe we could track adding that in a github issue in case it is useful for someone.

Hi @dnephin, thank you for the approval and suggestions. After briefly reviewing the CONNECT method on the http2 RFC, it seems like it might not support ping frames, but I have not had direct experience with proxies over http2 so I could be mistaken and will look into this further. I created a PR for WaitGroups and also added graceful shutdown of the client connection (which I only just noticed was not implemented- apologies for missing it earlier) https://github.com/hashicorp/consul/pull/9995. I hope that it is okay that I combined both the WaitGroup feature and the quick fix to gracefully close connections in the same PR- I am happy to separate them if necessary.
